Mathlib/LinearAlgebra/LinearProgramming.lean

```diff
@@ -0,0 +1,76 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak, Antoine Chambert-Loir
+-/
+import Mathlib.Analysis.Convex.Cone.Pointed
+import Mathlib.Algebra.Order.Group.Defs
+
+/-!
+
+# Linear programming
+
+TODO
+
+-/
+
+/-- Typically `M` is `ℝ^m` and `N` is `ℝ^n` -/
+structure LinearProgram (R V W M N : Type*) [OrderedRing R]
+    [AddCommGroup V] [Module R V] [AddTorsor V M]
+    [AddCommGroup W] [Module R W] [AddTorsor W N]
+    where
+  /-- Linear map -/
+  linmap : M →ᵃ[R] N
+  /-- Right-hand side -/
+  upper : N
```

Comment:
When I refactored it to not use `upper`, it ended up looking very strange:
```
/-- Typically `M` is `ℝ^m` and `N` is `ℝ^n` -/
structure LinearProgram (R V M N : Type*) [OrderedRing R]
    [AddCommGroup V] [Module R V] [AddTorsor V M]
    [AddCommGroup N] [Module R N]
    where
  /-- Affine map -/
  themap : M →ᵃ[R] N
  /-- Objective function -/
  objective : M →ᵃ[R] R
  /-- Cone defines nonnegative elements -/
  cone : PointedCone R N

variable {R V W M N : Type*} [OrderedRing R]
    [AddCommGroup V] [Module R V] [AddTorsor V M]
    [AddCommGroup N] [Module R N]

/-- `LP.primal = { x : M | LP.themap x ≤ LP.upper }` -/
def LinearProgram.primal (LP : LinearProgram R V M N) :=
  { x : M | -LP.themap x ∈ LP.cone }

/-- `LP.dual = { g : N →ᵃ[R] R | LP.objective = g ∘ LP.themap ∧ ∀ a, ∀ b, a ≤ b → g a ≤ g b }` -/
def LinearProgram.dual (LP : LinearProgram R V M N) :=
  { g : N →ᵃ[R] R | LP.objective = g ∘ LP.themap ∧
      ∀ a : N, ∀ b : N, (b -ᵥ a) ∈ LP.cone → g a ≤ g b }

-- From here on, we will need more assumptions than currently written

theorem LinearProgram.weakDuality (LP : LinearProgram R V M N)
    {c : M} (hc : c ∈ LP.primal) {d : N →ᵃ[R] R} (hd : d ∈ LP.dual) :
    LP.objective c ≤ d 0 := by
  simp_all [LinearProgram.primal, LinearProgram.dual]
```
Since we are maximizing in the primal, the feasibility condition ends up being `{ x : M | -LP.themap x ∈ LP.cone }`, which is ugly af.

Reply:
If you really want me to do it, I can flip the sign and proceed with this setting, but I am not happy about it, since we seemed to finally reached the consensus that Antoine Chambert-Loir's definition is the one we want to have in Mathlib.